### PR TITLE
Use circle 2.0

### PR DESCRIPTION
**Pull request by helenyuyu**
## What changes were proposed in this pull request?

Migrate to Circle 2.0

## How was this patch tested?
n/a

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/spark](https://github.com/palantir/spark), @helenyuyu! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            rsync --info=stats2,misc1,flist0 -a build_classes/ .
+          fi
+      - run: set -o pipefail && ./build/mvn -DskipTests -Phadoop-cloud -Phadoop-palantir -Pkinesis-asl -Pkubernetes -Pyarn -Phive -Psparkr install | tee -a "$CIRCLE_ARTIFACTS/mvn-install.log"
+      - run: >
+          rsync --info=stats2,misc1,flist0 -a --delete-excluded --prune-empty-dirs --exclude build_classes/ --exclude 'target/streams' --exclude 'assembly/target' --exclude 'common/network-yarn/target' --exclude 'examples/target' --exclude '***/*.jar' --include 'target/***' --include '**/' --exclude '*' . build_classes/
+      - run: |
+          # Make sbt fetch all the external deps to ~/.ivy2 so it gets cached
+          set -o pipefail && ./build/sbt -Phadoop-cloud -Phadoop-palantir -Pkinesis-asl -Pkubernetes -Pyarn -Phive -Psparkr externalDependencyClasspath | tee -a "$CIRCLE_ARTIFACTS/sbt-classpath.log"
+      - save_cache:
+          key: dependency-cache
```
Comment (dansanduleac) [inline .circleci/config.yml]: these keys are immutable, we need to collect the dependency list into a file then checksum it into the key

**Comment by robert3005** [comment]
Can we also get an image with python and r preinstalled?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      - test:
+          requires:
+            - build
+      - deploy:
+          requires:
+            - build
+            - test
+          filters:
+            branches:
+              only: master
```
Comment (helenyuyu) [inline .circleci/config.yml]: think these are AND'd, so its only run when the branch is master and the tag matches the regex. think we need two jobs here, with mutually exclusive conditions to hit both cases

File: .circleci/config.yml (hunk 1 above)
Comment (helenyuyu) [reply .circleci/config.yml]: is it enough to checksum the pom.xml?

**Comment by robert3005** [comment]
I think you need com/sun/jersey/api/container/filter/GZIPContentEncodingFilter for yarn tests
and org/apache/commons/logging/LogFactory generally

**Pull request merged**